### PR TITLE
Replace type_workarounds.NotImplementedType with types.NotImplementedType

### LINKED ISSUE #6775: Replace type_workarounds.NotImplementedType
**Description of the issue**

Python 3.10 (our minimum requirement) provides [types.NotImplementedType](https://docs.python.org/3.10/library/types.html?highlight=notimplementedtype#types.NotImplementedType).  The following workaround is redundant and can be replaced with that type -
https://github.com/quantumlib/Cirq/blob/351a08e52b7090cfab3a1ad07859ebf09d54052a/cirq-core/cirq/type_workarounds.py#L15-L23

**TODO**

- switch to types.NotImplementedType in the code
- delete the type_workarounds module

**Cirq version**

1.5.0.dev at 351a08e52b7090cfab3a1ad07859ebf09d54052a.

### PR DESCRIPTION
**Pull request by Ayobami-00**
- Replaced type_workarounds.NotImplementedType with types.NotImplementedType

- Updated code to use Python 3.10's built-in types.NotImplementedType, removing the need for the type_workarounds module.

- This change simplifies
the codebase by relying on standard types, allowing us to delete the now-redundant type_workarounds.py file.

Fixes #6775

### CODE REVIEW

File: cirq-core/cirq/circuits/circuit.py [hunk 1]
```diff
@@ -62,7 +62,7 @@
 from cirq.circuits.qasm_output import QasmOutput
 from cirq.circuits.text_diagram_drawer import TextDiagramDrawer
 from cirq.protocols import circuit_diagram_info_protocol
-from cirq.type_workarounds import NotImplementedType
+from types import NotImplementedType
```
Comment (pavoljuhas) [inline cirq-core/cirq/circuits/circuit.py:65]: Please move under the line 27 so this import is in the builtin-imports section and preserves their alphabetic order.
Same for the other sources below.

Comment (Ayobami-00) [reply cirq-core/cirq/circuits/circuit.py]: I've made the update.

I'll address the CI-reported errors now

Comment (Ayobami-00) [reply cirq-core/cirq/circuits/circuit.py]: @pavoljuhas I have addressed the CI-reported errors

**Comment by pavoljuhas** [comment]
For a faster testing consider setting up a Python development environment following
https://github.com/quantumlib/Cirq/blob/main/docs/dev/development.md#setting-up-an-environment

After that running `check/mypy` or `check/format-incremental` would do the CI checks locally.

**Comment by Ayobami-00** [comment]
@pavoljuhas, All checks have passed

**Review by pavoljuhas — APPROVED** [review]
Thank you for fixing this!

**Comment by Ayobami-00** [comment]
Thanks, @pavoljuhas, for reviewing and merging the PR!

**Pull request merged**